src/interval_sets.jl

```diff
@@ -0,0 +1,382 @@
+# which may change type during the union process)
+function Base.union(intervals::AbstractVector{T}) where T <: Interval
+    input = convert(Vector{T}, intervals)
+    if input === intervals
+        input = copy(intervals)
+    end
+    return union!(input)
+end
+
+"""
+    union!(intervals::AbstractVector{<:AbstractInterval})
+
+Flattens a vector of overlapping intervals in-place to be a smaller vector containing only
+non-overlapping intervals.
+"""
+function Base.union!(intervals::AbstractVector{<:AbstractInterval})
+    sort!(intervals)
+
+    i = 2
+    n = length(intervals)
+    while i <= n
+        prev = intervals[i - 1]
+        curr = intervals[i]
+
+        # If the current and previous intervals don't meet then move along
+        if !overlaps(prev, curr) && !contiguous(prev, curr)
+            i = i + 1
+
+        # If the two intervals meet then we absorb the current interval into
+        # the previous one.
+        else
+            intervals[i - 1] = merge(prev, curr)
+            deleteat!(intervals, i)
+            n -= 1
+        end
+    end
+
+    return intervals
+end
+
+"""
+    superset(intervals::AbstractArray{<:AbstractInterval}) -> Interval
+
+Create the smallest single interval which encompasses all of the provided intervals.
+"""
+function superset(intervals::AbstractArray{<:AbstractInterval})
+    left = minimum(LeftEndpoint.(intervals))
+    right = maximum(RightEndpoint.(intervals))
+
+    return Interval(left, right)
+end
+
+# set operations over multi-interval sets
+Base.intersect(x::AbstractIntervals, y::AbstractIntervals) = mergesets((inx, iny) -> inx && iny, x, y)
+Base.union(x::AbstractIntervals, y::AbstractIntervals) = mergesets((inx, iny) -> inx || iny, x, y)
+Base.setdiff(x::AbstractIntervals, y::AbstractIntervals) = mergesets((inx, iny) -> inx && !iny, x, y)
+Base.symdiff(x::AbstractIntervals, y::AbstractIntervals) = mergesets((inx, iny) -> inx ⊻ iny, x, y)
+Base.issubset(x::AbstractIntervals, y::AbstractIntervals) = isempty(setdiff(x, y))
+# may or may not be from Base (see top of `Intervals.jl`)
+Base.isdisjoint(x::AbstractIntervals, y::AbstractIntervals) = isempty(intersect(x, y))
+
+Base.in(x, y::AbstractVector{<:AbstractInterval}) = any(yᵢ -> x ∈ yᵢ, y)
+function Base.issetequal(x::AbstractIntervals, y::AbstractIntervals)
+    x, y, tracking = unbunch(union(x), union(y))
+    return x == y || (all(isempty, bunch(x, tracking)) && all(isempty, bunch(y, tracking)))
+end
+Base.length(x::AbstractInterval) = 1
+
+# order edges so that closed boundaries are on the outside: e.g. [( )]
+intersection_order(x::Endpoint) = isleft(x) ? !isclosed(x) : isclosed(x)
+intersection_isless_fn(::TrackStatically) = isless
+intersection_isless_fn(::TrackEachEndpoint) = function(x,y)
+    if isequal(x, y)
+        return isless(intersection_order(x), intersection_order(y))
+    else
+        return isless(x, y)
+    end
+end
```

Comment:
More inline with BlueStyle
```suggestion
function intersection_isless_fn(::TrackEachEndpoint)
    function (x,y)
        if isequal(x, y)
            return isless(intersection_order(x), intersection_order(y))
        else
            return isless(x, y)
        end
    end
end
```